src/main/kotlin/com/demonwav/mcdev/facet/MinecraftFacet.kt

```diff
@@ -0,0 +1,247 @@
+    private val roots: HashMultimap<SourceType, VirtualFile> = HashMultimap.create()
+
+    init {
+        configuration.facet = this
+    }
+
+    override fun initFacet() {
+        refresh()
+    }
+
+    fun refresh() {
+        // Don't allow parent types with child types in auto detected set
+        configuration.state.autoDetectTypes = PlatformType.removeParents(configuration.state.autoDetectTypes)
+
+        val userEnabled = configuration.state.userChosenTypes.entries.stream()
+            .filter { it.value }
+            .map { it.key }
+
+        val autoEnabled = configuration.state.autoDetectTypes.stream()
+            .filter { configuration.state.userChosenTypes[it] == null }
+
+        val allEnabled = Stream.concat(userEnabled, autoEnabled).collect(Collectors.toSet())
+
+        // Remove modules that aren't registered anymore
+        val toBeRemoved = modules.entries.stream()
+            .filter { !allEnabled.contains(it.key.platformType) }
+            .peek { it.value.dispose() }
+            .map { it.key }
+            .collect(Collectors.toSet())
+        toBeRemoved.forEach { modules.remove(it) }
+
+        // Add modules which are new
+        allEnabled.stream()
+            .map { it.type }
+            .filter { !modules.containsKey(it) }
+            .forEach { register(it) }
+
+        updateRoots()
+
+        ProjectView.getInstance(module.project).refresh()
+    }
+
+    private fun updateRoots() {
+        roots.clear()
+        val rootManager = ModuleRootManager.getInstance(module)
+        for (entry in rootManager.contentEntries) {
+            for (sourceFolder in entry.sourceFolders) {
+                when (sourceFolder.rootType) {
+                    JavaSourceRootType.SOURCE -> roots.put(SourceType.SOURCE, sourceFolder.file)
+                    JavaSourceRootType.TEST_SOURCE -> roots.put(SourceType.TEST_SOURCE, sourceFolder.file)
+                    JavaResourceRootType.RESOURCE -> roots.put(SourceType.RESOURCE, sourceFolder.file)
+                    JavaResourceRootType.TEST_RESOURCE -> roots.put(SourceType.TEST_RESOURCE, sourceFolder.file)
+                }
+            }
+        }
+    }
+
+    private fun register(type: AbstractModuleType<*>) {
+        type.performCreationSettingSetup(module.project)
+        modules[type] = type.generateModule(this)
+    }
+
+    @Contract(pure = true)
+    fun getModules(): Collection<AbstractModule> = modules.values
+    @Contract(pure = true)
+    fun getTypes(): Collection<AbstractModuleType<*>> = modules.keys
+
+    @Contract(value = "null -> false", pure = true)
+    fun isOfType(type: AbstractModuleType<*>?) = modules.containsKey(type)
+
+    @Contract(value = "null -> null", pure = true)
+    fun <T : AbstractModule> getModuleOfType(type: AbstractModuleType<T>?): T? {
+        @Suppress("UNCHECKED_CAST")
+        return modules[type as AbstractModuleType<*>] as? T
```

Comment:
It doesn't look like this fulfills the contract. Did you mean `type as AbstractModuleType<*>?` (the nullable type). In any case, an explicit null check before this would be probably better. Additionally, does it really even make sense to pass null to this method?